### ISSUE TITLE
Performance issues with new parser

### ISSUE THREAD

**Issue by pavarnos** [auraphp/Aura.Sql#142]
New parser is quite a bit slower than the old one...
<img width="1473" alt="transactions_-_new_relic" src="https://cloud.githubusercontent.com/assets/589595/22279254/79521fe8-e32e-11e6-8997-b4f7997b4c46.png">
<img width="1031" alt="_new_relic" src="https://cloud.githubusercontent.com/assets/589595/22279256/7c83b794-e32e-11e6-9a61-74d8d6c61c07.png">

**Comment by pmjones** [comment]
Yeesh, that's ugly. @jlacoude thoughts?

p.s. I figured it would be a noticeable difference, but not *that* noticeable. Maybe I'll put together a benchmark comparison. @pavarnos If you have a battery of query statements to parse, send them along.

**Comment by pmjones** [comment]
It's pretty significant. Benchmark code:

```
$statement = "SELECT :foo, :bar, :baz";
$values = [
    'foo' => 'foo',
    'bar' => 'bar',
    'baz' => ['baz1', 'baz2', 'baz3'],
];

$k = 10000;
$begin = microtime(true);
for ($i = 0; $i < $k; $i ++) {
    $pdo->prepareWithValues($statement, $values);
}
$elapsed = microtime(true) - $begin;
echo "Total: " . ($elapsed) . PHP_EOL;
```

Results:
```
$ php bench-2.x.php;php bench-2.x.php;php bench-2.x.php;php bench-2.x.php;php bench-2.x.php
Total: 0.24196696281433
Total: 0.25062203407288
Total: 0.25587511062622
Total: 0.24196290969849
Total: 0.24846506118774
$ php bench-3.x.php;php bench-3.x.php;php bench-3.x.php;php bench-3.x.php;php bench-3.x.php
Total: 0.84113097190857
Total: 0.78416991233826
Total: 0.77567601203918
Total: 0.78816509246826
Total: 0.87434482574463
$ 
```

That's about a 3x difference on a very short query string; I imagine it gets worse as the string gets longer.

**Comment by pmjones** [comment]
So, a series of questions:

- How important is it to automatically detect placeholders for array values, and expand them to comma-separated placeholders in the query string?

- If it's still important, how important is it to support non-utf-8 query strings?

- Is it possible to duplicate the character-by-character parsing wth a parser/rebuilder more like the older 2.x version, *and* avoid the bugs reported in #104 #107 #111 etc. ?

**Comment by pmjones** [comment]
Further, it may make sense to restrict the array-expansion to named placeholders only, which might help simplify the parser/rebuilder logic.

**Comment by pavarnos** [comment]
my 2c worth
- array values: essential. When i was hunting for a package a year ago, Aura was the only major one that did it that also had clean code. It was the deciding factor
- utf8 only? v3 is allowed BC breaks, so why not.
- would dropping support for ? placeholders simplify things? Aura.SqlQuery has some extra baggage around handling those for join clauses that could be cleaned out if so...

**Comment by pavarnos** [comment]
random idea (based on not understanding much about the parser or utf8 queries)...

would a naive approach work? eg since this is a new version and we are allowed BC breaks, make the parser do only 3 simple things
- strip comments
- the parser knows which placeholders are array values so can look for exactly `(:placeholder)` with a str_replace and do the array value substitution. Leave the rest for the PDO driver. Looking at docs/index.md, `WHERE...IN` is the only documented use case for array substitution? 
- look for tablename.fieldname with a regex and quote them. Could do a pre-scan of the query looking for ` FROM tablename`, ` JOIN (left|inner|outer|right) tablename`, ` AS alias`  etc to build a list of known tables / aliases to reduce the likelihood of altering quoted text values. Add documentation to let users know of this possibility and suggest that quoted text in sql should be bound parameters anyway (to make it portable / secure). The docs do not say anything about auto-quoting table and column names, so even this part could be dropped? Then the parser would only need to be called if an array parameter exists.

Would anything important break if we did this?

**Comment by JLacoude** [comment]
The performance problem comes from trying to extract the structure of the SQL string character per character.

As it's done mainly to handle the nested strings and comments I've tried to check if it could be done with regular expressions. It seems to be.

[Here is a method](https://gist.github.com/JLacoude/93beb8f56de233556fd7c028379cf5dc) which takes a SQL string as input and gives back an array of intervals containing the start, length, content and type of interval (comment, string or sql). I can not work on this more today but it should be possible to resolve lot of problems by applying the :placeholder transformation only in the sql intervals. We can keep the possibility to change the ? placeholder and maybe check for the ; character to still get multiple queries.

**Comment by harikt** [comment]
@JLacoude touch regx :( .

I think if we could not easily digest the regx, and if a bug occurs the time to patch it will be also heavy. Probably it need more experienced hands here at the time.

So the question is how much of these functionalities are really needed / used normally.

**Comment by pavarnos** [comment]
@JLacoude that is some awesome ninja regex-fu. 
Following @harikt  comments: would splitting it make sense? Run it in stages. 
1. strip comments
2. remove 2nd and subsequent queries (they are unused)
3. explode into chunks: quoted strings, sql code (keep the chunks in original string order)
4. replace array parameters in sql code
5. join(' ', $chunks)

?
would make it easier to test too...

**Comment by pmjones** [comment]
Good Lord, that's complex. Though maybe not more complex than the lexer, and maybe faster. @JLacoude try replacing the lexer with that, and incorporate array expansion if you can. (I'll experiment with it myself, too.)

**Comment by pmjones** [comment]
@pavarnos @JLacoude @harikt --

I have created a new branch, `re-rebuilder`, with the @JLacoude regex-fu instead of the @JLacoude lexer code, integrating the regex work into the older "Rebuilder" and ignoring the "Parser" classes. Benchmark results are:

```
$ php bench-2.x.php; php bench-2.x.php; php bench-2.x.php; php bench-2.x.php; php bench-2.x.php;
Total: 0.22703194618225
Total: 0.23022198677063
Total: 0.25027084350586
Total: 0.25984120368958
Total: 0.22256183624268
$ php bench-3.x.php; php bench-3.x.php; php bench-3.x.php; php bench-3.x.php; php bench-3.x.php;
Total: 0.41607689857483
Total: 0.41844487190247
Total: 0.44942402839661
Total: 0.46104001998901
Total: 0.4351909160614
```

So, better than the lexer (about 2x faster), but not as good as the original 2.x codebase (about half as fast).

Further thoughts?

**Comment by pmjones** [comment]
@pavarnos Would it be possible for you to send (perhaps by email) a long query string, and some fake values, that you are using in the application you showed the stats for above? Alternatively, would you be willing to try the `re-rebuilder` branch in your setup, and see how well it performs for you?

**Comment by harikt** [comment]
@pmjones this is my current thought. I am not in favor for regex. It is not for the regex is not powerful. But hard to understand ( at-least for me ).

I was again looking at https://github.com/auraphp/Aura.Sql/issues/104 and it do mention not to use regx and we are moving towards it.

Regarding your current rebuilder and regx .

I was just experimenting with this query string 

```
SELECT * FROM sometable
# Comment
/* Comment */
# Comment
-- Comment
WHERE id=4
# another comment
```

over https://regex101.com/r/gzWiFC/2 . It cannot understand certain comments . I haven't looked more closely what the real logic is behind the regx. Is that a bug with the regx or probably I didn't understand it as needed.

**Comment by pmjones** [comment]
@JLacoude Do you have any thoughts on the comments from @harikt above?

**Comment by JLacoude** [comment]
The logic of the regexp is to capture groups of comment and string. It takes everything until it finds the first character starting a string or comment. Then it captures it. Next try is started where this match ended thanks to the \G.

About your regex101 experiment:

I started with pgsql so '#' is not recognized as a comment with this regexp (yes we'll still need one parser per driver). So the first match is everything until it finds the first multi line comment and to the end of it (the pink part) which is put in the MLC group.

If you active the global flag to get the same thing you'll get with preg_match_all there is a second match with the single line comment.

**Comment by harikt** [comment]
Thank you for the explanations @JLacoude .

**Comment by pmjones** [comment]
@JLacoude I've been working with the new regex, and have a good level of understanding on how it works now. To compare it more closely to the 2.x rebuilder code, I pulled out the single/double line comment and the dollar-quoted string parts, and benchmarked it:

```
$ php bench-2.x.php; php bench-2.x.php; php bench-2.x.php; php bench-2.x.php; php bench-2.x.php;
Total: 0.24580502510071
Total: 0.23585891723633
Total: 0.24159097671509
Total: 0.24682283401489
Total: 0.23622798919678
$ php bench-3.x.php; php bench-3.x.php; php bench-3.x.php; php bench-3.x.php; php bench-3.x.php; 
Total: 0.32668781280518
Total: 0.31547093391418
Total: 0.31101989746094
Total: 0.32086491584778
Total: 0.31089305877686
$ 
```

So, for the exact same functionality as the old 2.x rebuilder, the new regex-fu approach takes about 25% longer.

I've been arguing with myself on two different sides since #104 was raised. Side 1 is "pretty good performance for the majority of users and cases" and side 2 is "cover all the imaginable cases even though infrequent for most users".

After exercising the lexer approach, and now the regex-fu approach, and given the amount of time we've spent on them, I have come to these conclusions:

1. Go back to something more like old 2.x rebuilder approach, and make allowance within it to handle double-colons and dollar-quoted strings, **but** continue to not-handle single- and double-line comments. That means if there are comments with things that look like `:foo` and `?` placeholders, the query might get messed up.

2. However, make the rebuilder injectable or settable, so that you @JLacoude (and others) can put together customer rebuilders for your own cases, to handle comments and other things as you wish.

@JLacoude you've done a hell of a lot of work here, and it's been worth doing if only to discover the boundaries and limitations of the different approaches. I thought the performance implications would be less dramatic than they have been, but as it is, the performance tradeoff is hard for me to take.

Are there any further comments or arguments to be made that might affect those conclusions?

**Comment by pavarnos** [comment]
@pmjones emailed example queries to your gmail address as requested.

i like the idea of pluggable / composable / injectable rebuilders so that for a specific project if you need dollar quoted strings you can have them. And if you don't, you don't need to pay the performance cost. Taking it to ridiculous extremes: do you have any data on how important array parameter expansion is to people who use this component? eg should it ship with a NullRebuilder by default?

**Comment by pmjones** [comment]
Received!

Yeah, settable rebuilder/parser, a la the @JLacoude work, is definitely the way to go.

**Comment by harikt** [comment]
@pmjones we could add all these Rebuilders, and mention about those in the docs. The users can choose from various implementations may be nice.

**Comment by JLacoude** [comment]
@pmjones I'm all for settable parsers

@pavarnos I don't know about the use of array parameters but the functionality I know I use a lot is the fact you can use the same named placeholder multiple times in a statement which PDO does not allow.

**Comment by pavarnos** [comment]
@JLacoude i use (same named placeholder multiple times) heavily too. 

http://php.net/manual/en/pdo.prepare.php has some interesting comments that may save some code / time (emphasis mine):

> You must include a **unique** parameter marker for **each value** you wish to pass in to the statement when you call PDOStatement::execute(). You cannot use a named parameter marker of the same name more than once in a prepared statement, unless emulation mode is on.

MySQL emulates by default, which is possibly why this has never been an issue for me? 

> You cannot use both named and question mark parameter markers within the same SQL statement; pick one or the other parameter style

Aura\Sql currently allows these to be mixed, but at cost of some complexity. Is this a feature worth carrying forward?

> PDO will emulate prepared statements/bound parameters for drivers that do not natively support them, and can also **rewrite named or question mark style parameter markers to something more appropriate**, if the driver supports one style but not the other.

Does the rebuilder need to spend a lot of effort doing it by default if the driver can do it faster? Perhaps an optional QuestionMarkToNamedParameterRebuilder might be useful to some people?

**Comment by pavarnos** [comment]
Have been doing some more performance testing the last few days. Cut down the Aura DecoratedPDO and put the naive parser into it. No performance difference.

Have just switched to the re-parser branch. Will monitor that for a day or so to see how performance goes.

**Comment by pavarnos** [comment]
No discernable difference in performance for re-parser branch.
<img width="1039" alt="newrelic" src="https://cloud.githubusercontent.com/assets/589595/23001848/156f8fb8-f44c-11e6-8254-1dc959da5cfb.png">
Tested over a shorter time period, and site is not so busy at the moment, but if there was an issue we should have seen it.
Looks like you are good to go: OK to close this issue.

**Comment by pmjones** [comment]
Thanks everyone for the analysis and critique here. This issue is now fixed by #144.
